S3Client/S3Client.cs

```diff
@@ -0,0 +1,132 @@
+      {
+        BucketName = _config.BucketName,
+        Prefix = _config.Prefix,
+      };
+
+      var response = await _client.ListObjectsV2Async(listRequest);
+      return response.S3Objects.Select(obj => new S3Image
+      {
+        Key = obj.Key,
+        Link = $"{GetBaseUrl()}/{obj.Key}",
+      }).ToArray();
+    }
+
+    public async Task<S3Image> UploadImage(string hash, string dataB64)
+    {
+      var key = $"{_config.Prefix}/{hash}";
+      var imageStream = new System.IO.MemoryStream(Convert.FromBase64String(dataB64));
+
+      // Get content type of image for S3
+      string contentType = "image/unknown";
+      using (var image = Image.FromStream(imageStream))
+      {
+        Guid guid = image.RawFormat.Guid;
+        foreach (ImageCodecInfo codec in ImageCodecInfo.GetImageDecoders())
+        {
+          if (codec.FormatID == guid)
+          {
+            contentType = codec.MimeType;
+            break;
+          }
+        }
+      }
+
+      var putRequest = new PutObjectRequest
+      {
+        BucketName = _config.BucketName,
+        Key = $"{_config.Prefix}/{hash}",
+        ContentType = contentType,
+        InputStream = imageStream,
+        CannedACL = S3CannedACL.PublicRead,
+      };
+
+      var response = await _client.PutObjectAsync(putRequest);
+      if (response.HttpStatusCode == System.Net.HttpStatusCode.OK)
+      {
+        return new S3Image
+        {
+          Key = key,
+          Link = $"{GetBaseUrl()}/{key}",
+        };
+      } else
+      {
+        return null;
+      }
+    }
+
+    public async Task<bool> TestConnection()
+    {
+      // Check if an S3 request can be successfully made
+      try
+      {
+        var listRequest = new ListObjectsV2Request
+        {
+          BucketName = _config.BucketName,
+          Prefix = _config.Prefix,
+        };
+
+        var response = await _client.ListObjectsV2Async(listRequest);
+        return response.HttpStatusCode == System.Net.HttpStatusCode.OK;
+      }
+      catch
+      {
+        return false;
+      }
+    }
+
+    public void Dispose()
+    {
+      _client?.Dispose();
+      GC.SuppressFinalize(this);
+    }
+
+    private string GetBaseUrl()
+    {
+      string url = _config.CustomDomain ?? _config.Endpoint;
+
+      // I was considering removing the url setting above entirely, as it doesn't seem to work. But it may just be my problem, so if it fails, continue to setting it through getting the bucket
+      if (string.IsNullOrEmpty(url))
+      {
+        // Fetch the bucket region
+        var regionResponse = _client.GetBucketLocationAsync(new GetBucketLocationRequest
+        {
+          BucketName = _config.BucketName
+        }).GetAwaiter().GetResult();
+
+        var region = regionResponse.Location?.Value ?? "us-east-1"; // Default to us-east-1 if region is null
+        url = $"https://{_config.BucketName}.s3.{region}.amazonaws.com";
+      }
+
+      return url;
```

Comment:
This should resolve the problem you were initially encountering. It also eliminates the need for querying the region of the bucket from S3. The bucket name needs to be prefixed to the endpoint's host for the [virtual-host-style URLs](https://docs.aws.amazon.com/AmazonS3/latest/userguide/VirtualHosting.html#virtual-hosted-style-access). This should work with the [global endpoint](https://docs.aws.amazon.com/AmazonS3/latest/userguide/VirtualHosting.html#deprecated-global-endpoint) `s3.amazonaws.com` and [regional endpoints](https://docs.aws.amazon.com/general/latest/gr/s3.html) (`s3.us-east-1.amazonaws.com`, etc.) This does make assumptions that other S3 stacks ([MinIO](https://min.io/), etc.) work similar, but `CustomDomain` should provide flexibility there in the event the default bucket name + endpoint value doesn't work.

```suggestion
      if (!string.IsNullOrEmpty(_config.CustomDomain))
      {
        return _config.CustomDomain;
      }

      // Prefix the bucket name to the endpoint's host
      var endpoint = new UriBuilder(_config.Endpoint);
      endpoint.Host = $"{_config.BucketName}.${endpoint.Host}";
      return endpoint.Uri.ToString();
```

Reply:
Will try this out later.

Reply:
This is almost working, but not quite. For one, the $ you left breaks things, but secondly, there is still no bucket region. I will see about getting that in.

Reply:
I'm going to have to do some pretty uh, weird things to get this to work. See, the problem is that while the endpoint is `s3.amazonaws.com`, and that's all well and good, for my bucket I'm going to need `discordbee-suflors.s3.us-west-1.amazonaws.com`, as you can see I need something in between the `s3` and `amazonaws` bit. I could see if I can just put the whole endpoint as `s3.us-west-1.amazonaws.com`, I have no idea if this would get in the way of any of the other code, but I will check.